run.go

```diff
@@ -8,31 +8,41 @@ import (
 // Start will launch the go routines in the background to perform Autopilot.
 // When the context passed in is cancelled or the Stop method is called
 // then these routines will exit.
-func (a *Autopilot) Start(ctx context.Context) {
+func (a *Autopilot) Start(ctx context.Context) error {
 	a.runLock.Lock()
 	defer a.runLock.Unlock()
 
 	// already running so there is nothing to do
-	if a.running {
-		return
+	if a.execution != nil && a.execution.status == Running {
+		return nil
 	}
 
 	ctx, shutdown := context.WithCancel(ctx)
-	a.shutdown = shutdown
 	a.startTime = a.time.Now()
-	a.done = make(chan struct{})
-
-	// While a go routine executed by a.run below will periodically
-	// update the state, we want to go ahead and force updating it now
-	// so that during a leadership transfer we don't report an empty
-	// autopilot state. We put a pretty small timeout on this though
-	// so as to prevent leader establishment from taking too long
-	updateCtx, updateCancel := context.WithTimeout(ctx, time.Second)
-	defer updateCancel()
-	a.updateState(updateCtx)
-
-	go a.run(ctx)
-	a.running = true
+
+	exec := &execInfo{
+		status:   Running,
+		shutdown: shutdown,
+		done:     make(chan struct{}),
+	}
+
+	if a.execution == nil || a.execution.status == NotRunning {
+		// While a go routine executed by a.run below will periodically
+		// update the state, we want to go ahead and force updating it now
+		// so that during a leadership transfer we don't report an empty
+		// autopilot state. We put a pretty small timeout on this though
+		// so as to prevent leader establishment from taking too long. This
+		// only is done if we are not running and not shutting down so
+		// as to prevent conflicts with any autopilot routine just finishing
+		// up such as when restarting autopilot.
+		updateCtx, updateCancel := context.WithTimeout(ctx, time.Second)
+		defer updateCancel()
+		a.updateState(updateCtx)
+	}
+
+	go a.run(ctx, exec)
+	a.execution = exec
```

Comment:
I wonder if we need to consider the case where `run()` exits from a context cancellation here?

If I have 2 autopilot `Starts()` with different contexts, and the second one is canceled, we will end up overwriting a currently running execution.

```
ctx1, _ := context.WithTimeout(ctx, time.Minute)
ctx2, _ := context.WithTimeout(ctx, time.Second)

ap.Start(ctx1)
ap.Start(ctx2)

time.Sleep(5*time.Second)

// status will be NotRunning here, but ctx1's goroutines are still active.
status, _ := ap. IsRunning()
```

This is a contrived example of course, and we still need `run()` to be spun off into a goroutine, but perhaps we should wait to set `a.execution = exec` until we are sure that we are actually running autopilot?

Reply:
Ah well I realized my example does not actually do what I thought, because of the `runLock` controlling access. 🗝️ 

But my questions still stands, should we wait to set a.execution = exec until we are sure that we are actually running autopilot?

Reply:
We can't because if you called `Start` followed by another `Start` there would be a race between the go routine started by the first call trying to get to a place where it has updated the execution and where the second start checks if it is already running.